File: July2022/Java/MaximumAreaOfAPieceOfCakeAfterHorizontalAndVerticalCuts.java

// @saorav21994
// TC : O(n)
// SC : O(1)

// Multiple companie(s) including Amazon ask this in online round.

class Solution {
    public int maxArea(int h, int w, int[] horizontalCuts, int[] verticalCuts) {
        int hl = horizontalCuts.length;
        int vl = verticalCuts.length;
        Arrays.sort(horizontalCuts);
        Arrays.sort(verticalCuts);
        int maxh = 0, maxv = 0;
        int prev = 0;
        for (int i = 0; i < hl; i++) {
            maxh = Math.max(maxh, horizontalCuts[i] - prev);
            prev = horizontalCuts[i];
        }
        maxh = Math.max(h-prev, maxh);
        // System.out.println("maxh = " + maxh);
        prev = 0;
        for (int i = 0; i < vl; i++) {
            maxv = Math.max(maxv, verticalCuts[i] - prev);
            prev = verticalCuts[i];
        }
        maxv = Math.max(w-prev, maxv);
        // System.out.println("maxv = " + maxv);
        long res = (long)((long)((long)maxh%1000000007 * (long)maxv%1000000007))%1000000007;
        return (int)res;
    }
}

File: July2022/C++/MaximumAreaOfAPieceOfCakeAfterHorizontalAndVerticalCuts.cpp
class Solution {
public:
    // TC: O(vlogv) + O(hlogh) => v = no. of vertical cuts, h -> no. of horizontal cuts
    // SC: O(1)
    int maxArea(int h, int w, vector<int>& horizontalCuts, vector<int>& verticalCuts) {
        sort(horizontalCuts.begin(), horizontalCuts.end());
        sort(verticalCuts.begin(), verticalCuts.end());
        
        int hLength = horizontalCuts.size();
        int vLength = verticalCuts.size();
        
        int maxHeight = max(horizontalCuts[0] - 0, h - horizontalCuts[hLength - 1]);
        for (int i = 1; i < hLength; i++) {
            int currHeight = horizontalCuts[i] - horizontalCuts[i - 1];
            maxHeight = max(maxHeight, currHeight);
        }
        
        int maxWidth = max(verticalCuts[0] - 0, w - verticalCuts[vLength - 1]);
        for (int i = 1; i < vLength; i++) {
            int currWidth = verticalCuts[i] - verticalCuts[i - 1];
            maxWidth = max(maxWidth, currWidth);
        }
        
        return (long) maxHeight * maxWidth % 1000000007;
    }
};
